com.sap.cloud.lm.sl.cf.process/src/main/java/com/sap/cloud/lm/sl/cf/process/steps/PollServiceCreateOrUpdateOperationsExecution.java

```diff
@@ -1,92 +1,128 @@
-        CloudServiceExtended service) {
-        Map<String, Object> cloudServiceInstance = serviceOperationExecutor.executeServiceOperation(service,
-            () -> serviceInstanceGetter.getServiceInstanceEntity(client, service.getName(), StepsUtil.getSpaceId(execution.getContext())),
-            execution.getStepLogger());
-
-        if (cloudServiceInstance == null || cloudServiceInstance.isEmpty()) {
-            handleMissingServiceInstance(execution, service);
-            return null;
+    protected ServiceOperation mapOperationState(StepLogger stepLogger, ServiceOperation lastServiceOperation,
+                                                 CloudServiceExtended service) {
+        lastServiceOperation = super.mapOperationState(stepLogger, lastServiceOperation, service);
+        // Be fault tolerant on failure on update of service
+        if (lastServiceOperation.getType() == ServiceOperationType.UPDATE
+            && lastServiceOperation.getState() == ServiceOperationState.FAILED) {
+            stepLogger.warn(Messages.FAILED_SERVICE_UPDATE, service.getName(), lastServiceOperation.getDescription());
+            return new ServiceOperation(lastServiceOperation.getType(),
+                                        lastServiceOperation.getDescription(),
+                                        ServiceOperationState.SUCCEEDED);
         }
-        return getLastOperation(execution, cloudServiceInstance);
+        return lastServiceOperation;
     }
 
-    @SuppressWarnings("unchecked")
-    private ServiceOperation getLastOperation(ExecutionWrapper execution, Map<String, Object> cloudServiceInstance) {
-        Map<String, Object> lastOperationAsMap = (Map<String, Object>) cloudServiceInstance.get(LAST_SERVICE_OPERATION);
-        ServiceOperation lastOperation = parseServiceOperationFromMap(lastOperationAsMap);
-        // TODO Separate create and update steps
-        // Be fault tolerant on failure on update of service
-        if (lastOperation.getType() == ServiceOperationType.UPDATE && lastOperation.getState() == ServiceOperationState.FAILED) {
+    @Override
+    protected void handleMissingOperationState(StepLogger stepLogger, CloudServiceExtended service) {
+        if (!service.isOptional()) {
+            throw new SLException(Messages.CANNOT_RETRIEVE_INSTANCE_OF_SERVICE, service.getName());
+        }
+        // Here we're assuming that we cannot retrieve the service instance, because its creation was synchronous and it failed. If that
+        // is really the case, then showing a warning progress message to the user is unnecessary, since one should have been shown back
+        // in CreateOrUpdateServicesStep.
+        stepLogger.warnWithoutProgressMessage(Messages.CANNOT_RETRIEVE_SERVICE_INSTANCE_OF_OPTIONAL_SERVICE, service.getName());
+    }
+
+    @Override
+    protected void reportServiceState(ExecutionWrapper execution, CloudServiceExtended service, ServiceOperation lastServiceOperation) {
```

Comment:
It is strange to have this method here because you already have the serviceProgressReporter ?